File: 087/eu087.go
// project euler (projecteuler.net) problem 87
// solution by Kevin Retzke (retzkek@gmail.com), May 2012
package main

import (
	"fmt"
	"math"
)

type Primes struct {
	Primes []int
	Last   int
}

// Init initializes a Primes struct with the first two primes.
func (p *Primes) Init() {
	p.Primes = []int{2, 3}
	p.Last = 3
}

// Next computes, returns, and appends the next prime number.
func (p *Primes) Next() int {
	next := 0
	i := p.Last + 2
	for next == 0 {
		sqrti := math.Sqrt(float64(i))
		isPrime := true
		for _, p := range p.Primes {
			if i%p == 0 {
				isPrime = false
				i += 2
				break
			}
			if float64(p) > sqrti {
				break
			}
		}
		if isPrime {
			next = i
		}
	}
	p.Primes = append(p.Primes, next)
	p.Last = next
	return next
}

func main() {
	p := new(Primes)
	p.Init()
	for p.Last*p.Last < 50e6 {
		p.Next()
	}
    p2 := make([]int, len(p.Primes))
    p3 := make([]int, len(p.Primes))
    p4 := make([]int, len(p.Primes))
    for i,p := range p.Primes {
        p2[i] = p*p
        p3[i] = p2[i]*p
        p4[i] = p3[i]*p
    }
    sumset := make(map[int] bool)
    for _,a := range p2 {
        for _,b := range p3 {
            if b > 50e6 {
                break
            }
            for _,c := range p4 {
                if c > 50e6 {
                    break
                }
                sum := a+b+c
                if sum < 50e6 {
                    sumset[sum] = true
                }
            }
        }
    }
	fmt.Println(len(sumset))
}

File: 087/eu087.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problem 87
solution by Kevin Retzke, May 2012
"""
import math

class primes:
    """
    Calculates prime numbers sequentially.

    METHODS

    next()      returns the next prime
    """
    def __init__(self):
        self.primes = [2,3]
        #self.primeset = set(self.primes)
        self.lastPrime = 3
    def next(self):
        """
        calculate and return next prime
        """
        nextPrime = None
        i = self.lastPrime+2
        while nextPrime is None:
            sqrt_i = math.sqrt(i)
            isPrime = True
            for p in self.primes:
                if i%p == 0:
                    isPrime = False
                    i += 2
                    break
                if p > sqrt_i:
                    break
            if isPrime:
                nextPrime = i
        self.primes.append(nextPrime)
        #self.primeset.add(nextPrime)
        self.lastPrime = nextPrime
        return nextPrime

if __name__ == '__main__':
    p = primes()
    while p.lastPrime**2 < 50e6:
        p.next()
    sumset = set()
    for p2 in [n*n for n in p.primes if n*n < 50e6]:
        for p3 in [n**3 for n in p.primes if n**3 < 50e6]:
            for p4 in [n**4 for n in p.primes if n**4 < 50e6]:
                n = p2+p3+p4
                if n < 50e6:
                    sumset.add(p2+p3+p4)
    print len(sumset)
